spec/02-integration/05-proxy/02-upstream_headers_spec.lua

```diff
@@ -0,0 +1,786 @@
+        }
+
+        assert.equal("127.0.0.1", headers["x-real-ip"])
+      end)
+
+      it("should be replaced if present in request", function()
+        local headers = request_headers {
+          ["Host"]      = "headers-inspect.com",
+          ["X-Real-IP"] = "10.0.0.1",
+        }
+
+        assert.equal("127.0.0.1", headers["x-real-ip"])
+      end)
+    end)
+
+    describe("X-Forwarded-For", function()
+      it("should be added if not present in request", function()
+        local headers = request_headers {
+          ["Host"] = "headers-inspect.com",
+        }
+
+        assert.equal("127.0.0.1", headers["x-forwarded-for"])
+      end)
+
+      it("should be appended if present in request", function()
+        local headers = request_headers {
+          ["Host"]            = "headers-inspect.com",
+          ["X-Forwarded-For"] = "10.0.0.1",
+        }
+
+        assert.equal("10.0.0.1, 127.0.0.1", headers["x-forwarded-for"])
+      end)
+    end)
+
+    describe("X-Forwarded-Proto", function()
+      it("should be added if not present in request", function()
+        local headers = request_headers {
+          ["Host"] = "headers-inspect.com",
+        }
+
+        assert.equal("http", headers["x-forwarded-proto"])
+      end)
+
+      it("should be replaced if present in request", function()
+        local headers = request_headers {
+          ["Host"]              = "headers-inspect.com",
+          ["X-Forwarded-Proto"] = "https",
+        }
+
+        assert.equal("http", headers["x-forwarded-proto"])
+      end)
+    end)
+
+    describe("X-Forwarded-Host", function()
+      it("should be added if not present in request", function()
+        local headers = request_headers {
+          ["Host"] = "headers-inspect.com",
+        }
+
+        assert.equal("headers-inspect.com", headers["x-forwarded-host"])
+      end)
+
+      it("should be replaced if present in request", function()
+        local headers = request_headers {
+          ["Host"]             = "headers-inspect.com",
+          ["X-Forwarded-Host"] = "example.com",
+        }
+
+        assert.equal("headers-inspect.com", headers["x-forwarded-host"])
+      end)
+    end)
+
+    describe("X-Forwarded-Port", function()
+      it("should be added if not present in request", function()
+        local headers = request_headers {
+          ["Host"] = "headers-inspect.com",
+        }
+
+        assert.equal(helpers.test_conf.proxy_port, tonumber(headers["x-forwarded-port"]))
+      end)
+
+      it("should be replaced if present in request", function()
+        local headers = request_headers {
+          ["Host"]             = "headers-inspect.com",
+          ["X-Forwarded-Port"] = "80",
+        }
+
+        assert.equal(helpers.test_conf.proxy_port, tonumber(headers["x-forwarded-port"]))
+      end)
+    end)
+
+    describe("with the downstream host preserved (pending: because the test fails to connect to API when preserve_host = true)", function()
```

Comment:
Maybe we can remove this "pending" comment now? Assuming this was related to one of the recent fixes on master